### ISSUE TITLE
remainder operators - % and mod and rem

### ISSUE THREAD

**Issue by damianmoz** [chapel-lang/chapel#19255]
The remainder operation is defined as
```chapel
    remainder(x, y, mode) = x - rounded(x / y, mode) * y            (1)
```
where rounded returns the rounded result of the division (x / y) and mode is one of the IEEE 754 rounding modes:

a)      ceiling'd (or whatever it is called)
b)      floored
c)      truncated
d)      rounded to nearest, with ties (0.5) mapping to the adjacent integer which is even
e)      rounded to nearest, with ties (0.5) mapping to the adjacent integer with the larger magnitude (i.e. away from zero)

Chapel does (b) for all types. But unlike C/C++ which does (c) for all types, Chapel only does it for integral types. And while (d) for floating point types exists in C/C++ as mandated by the IEEE 754 standard, that functionality is missing in Chapel.

This issue is the start of the walk down the road to address these shortcomings.

Before going further, the rounded function above is not the same as the Chapel 'round' function, although the Chapel round function is this generic rounded function if the last mode is used.

It is counterproduction to discuss or argue about the mathematical correctness of Eq(1). Some disciplines use one form of a remainder functionality to the exclusion of all others. And some disciplines demand that the remainder operation satisfy more, or less, relationships, or identities, than others. Either way, the remainder operation is defined as above in Eq(1) consistently across those modes and needs to exist for at least (b), (c) and (d) cases. I have never seen a need for (a) or (e) remainders so I have put these as very **low** priority. Please correct me if that is flawed.

One thing which highlights the mish-mash among them all is the name used by various programming languages. In fact some languages call an apple an orange and vica versa.

The various rounding modes and the names used for the remainder operator in various languages for those modes follow. Note that some languages use an operator for this operation, others use a function.

a)      ceiling`d rounding - very low priority

b)      floored rounded
        - `mod` in Chapel (function)
        - `mod` in Ada (operator)
        - `mod` in Lisp (operator)
        - `mod` in Julia (??)
        - `modulo` Fortran (routine)
        - `mod` in MATLAB
        - `mod` in ....
        - **missing** in C/C++

c)      truncated rounding
        - `%` in Chapel for integral types
        - `%` in Swift for integral types
        - `mod` in Fortran for ALL types (function)
        - `%`/`fmod` in C for integral/floating types (operator/function)
        - `rem` in Ada for integers (operator)
        - `rem` in LISP for ALL types (operator)
        - `rem` in MATLAB
        - `%` in Julia

d)      rounded to nearest, ties (0.5) mapping to the adjacent integer which is even
        - `rem` in C for floating types (function)
        - `rem` in Swift for floating types 
        - often languages only define this for real arguments 

e)      rounded to away, ties (0.5) to mapping to the adjacent integer which has the largest magnitude - very low priority

Based largely on what Chapel does already because I do not want to break anything and using operator notation for now. So, in summary we have
```chapel
    x mod y ... is ... remainder(x, y, floor)
    x % y ..... is ... remainder(x, y, truncated)
    x rem y ... is ... remainder(x, y, round to nearest with ties to even)
```
That choice also means that there is no strict conflict with C/C++ and it follows those satisfying IEEE 754 for **rem**.

**const** task[1..5] = .......

Task#1:

The `%` operator in Chapel needs to made generic and accept real(w) numbers, for now.

Task#2:

Make a decision:

a)  do we push for a `mod` operator and a `rem` operator, or
b)  do we leave `mod` as a routine and add a `rem` routine.

With '%' as an operator, leaving the others as a routine is inconsistent.

With the other two as operators, we also need `rem=` and `mod=` operators.

Task#3:

Integrate the `rem` operator or routine for real(w) types.

Task#4:

Integrate the `rem` operator or routine for int(w) types.

Task#5:

Enhance `mod` and `rem` to handle complex(w) numbers if that makes sense.

**Comment by bradcray** [comment]
I just stumbled back across this and it resonated with questions @jeremiah-corrado has been asking recently, so wanted to tag him on it.

**Comment by damianmoz** [comment]
Let me know if anything is unclear or needs a better explanation.

**Comment by jeremiah-corrado** [comment]
Thank you for the very detailed explanation! 

This does relate somewhat to a recent discussion about the behavior of the `%` operator and `mod` function for the `bigint` type. ([Here](https://github.com/chapel-lang/chapel/issues/17713#issuecomment-1158940788) we were talking about ensuring that the rounding behavior for `%` is **truncating (c)** on integral and bigint types, and that `mod` is **flooring (b)** for both integral and bigint.)

There was also some discussion of including the `rem` (round to nearest (d)) behavior for bigint if it's something that users are looking for (and since this is the definition of modulo in the underlying [GMP](https://chapel-lang.org/docs/modules/standard/GMP.html) implementation).

We had not discussed extending these operators and functions to support `real`s; however I am open to the idea since there appears to be a precedent in other languages. 

For complex numbers, I am a little uncertain. Is this something that you've seen support for elsewhere?

**Comment by damianmoz** [comment]
What is your context please? I was not discussing things in a BigInt context, just an IIEEE 754 context. If the latter, you do not get any choice if Chapel is claiming IEEE 754 compliance.

As for your last question, the answer is No, although that is strictly Not Yet. That is actually part of a bigger question. Walk before you run.

**Comment by damianmoz** [comment]
Have you read the IEEE 754 standard cover to cover? And the floating point features in the C standard similarly?

**Comment by jeremiah-corrado** [comment]
> What is your context please? I was not discussing things in a BigInt context, just an IIEEE 754 context

Sorry for the confusion! I was just relaying some recent discussion on the topic of mod operators in Chapel since Brad suggested that these topics are sort of related. This was just in case you were curious about how some of the team members are thinking about the topic in general. 

I was also pointing out that, as far as I know, our analysis/rethinking of modulo behavior in chapel has not made its way towards floats yet. But as you suggest, this is certainly something worth discussing since it's a part of the IEEE 754 standard (which I have not read cover to cover).

**Comment by damianmoz** [comment]
Thanks heaps for the reference. Interesting and insightful reading.

You are probably better off using the term `floating point number` rather than `float`, even if the former is a bit wordy. In some contexts, a `float` too easily can be construed to imply a C/C++ 32-bit floating point number
